zookeeper-server/src/main/java/org/apache/zookeeper/server/persistence/FileTxnSnapLog.java

```diff
@@ -232,11 +242,19 @@ public long restore(DataTree dt, Map<Long, Integer> sessions, PlayBackListener l
         } else {
             trustEmptyDB = autoCreateDB;
         }
+
         if (-1L == deserializeResult) {
             /* this means that we couldn't find any snapshot, so we need to
              * initialize an empty database (reported in ZOOKEEPER-2325) */
             if (txnLog.getLastLoggedZxid() != -1) {
-                throw new IOException("No snapshot found, but there are log entries. " + "Something is broken!");
+                // ZOOKEEPER-3056: provides an escape hatch for users upgrading
+                // from old versions of zookeeper (3.4.x, pre 3.5.3).
+                if (!trustEmptySnapshot) {
+                    throw new IOException(EMPTY_SNAPSHOT_WARNING + "Something is broken!");
+                } else {
+                    LOG.warn(EMPTY_SNAPSHOT_WARNING + "This should only be allowed during upgrading.");
+                    trustEmptySnapshot = false;
```

Comment:
I am not an expert of this part, just want to share my doubts

Ideally we will hit the case of the issue only while the server is starting up, not during internal reloads of the database, it is allowed to have such lack of snapshot file only during bootstrap.
For instance I see that this method ("restore") is called in QuorumPeer#getLastLoggedZxid()
if we refactor the code in the future and change the boot sequence this change won't take effect.


I wonder if there is a cleaner way to allow this special case only during "start" and not during all of the other code paths that leads to the invocations of this method.

Reply:
>> during internal reloads of the database

database will only be loaded once from disk during a process's life time now we don't clear the in memory data base anymore every time a server (object) restarts; so this is less a concern. 

>> I wonder if there is a cleaner way to allow this special case only during "start" and not during all of the other code paths

"start" and "other code paths" are no different, from server point of view they are all recovery cases, be it a first start up or recovery from faults. So if we care about empty snapshot, we should check in all places. If we ignore empty snapshot, we should ignore in all places. The check logic and the skip check logic should be co-located, so refactoring sequence of start ups will not impact that.

Reply:
If we call this method only once.
Why should we reset this variable, as it won't be used anymore?

Can you please give an example of multiple calls of this method?

Reply:
>> we don't clear the in memory data base anymore

Actually, we still clear database sometimes (the case I was referring to was the database retain introduced in ZOOKEEPER-2678 and following up jira). There are two cases we still clear database: a recovering server get a TRUNC from leader; or a recovering server get a SNAP sync from leader. Both cases will clear db, leads to reload local snap to initialize the database.

In any case, one invariant here is the first invoke of `restore` will always happen when a fresh server is starting. I don't think a future refactoring will change this fact (if, there will be such refactoring at all) given the essence of this step during start up. With this invariant, we need reset the flag so further database reloading at run time does not accidentally skip the snapshot validation.

Now, one alternative is instead of trying to be clever and reset the flag in code, we remove the reset, but require ops admin to restart the zookeeper process after upgrading by reset the flag manually. This means upgrading a cluster will trigger leader election at least twice, instead of once, which seems not too bad, with the benefit of making code easier to understand. Even in this case, we can still keep this reset logic in code as an extra layer of protection, as that logic will only introduce false negative instead of false positive in terms of the data safety of the system, in worst case.

Reply:
I am okay with keeping this reset logic.

In fact I found those other execution path that lead to calls to *restore*
and I asked for confirmation.

Reply:
I'm tempted toward the unclever solution of not resetting the flag since this is easy to reason about. Unless we can guarantee that a snapshot is created by the server as the flag is flipped then it makes sense to have the admin bare the risk of operating in "snapshot not required" mode.

Reply:
removed reset logic and update doc to recommend an advisory flag reset after upgrade.